homeworks/Dmytro.Tyshchenko_DmytroTy/homework_6/task2.js

```diff
@@ -0,0 +1,84 @@
+const horsesList = [['Big Racket', 'Thoroughbred'], ['Shareef Dancer', 'Thoroughbred'],
+    ['Green Monkey', 'Thoroughbred'], ['Annihilator', 'Thoroughbred'],
+    ['Seattle Dancer', 'Thoroughbred'], ['Mystic Park', 'Standardbred'],
+    ['Pine Chip', 'Standardbred'], ['Sardar\t', 'Thoroughbred'],
+    ['Poetin\t', 'Brandenburger'], ['Lord Sinclair', 'Bavarian Warmblood']];
+
+class Horse {
+    constructor(name, breed) {
+        this.name = name;
+        this.breed = breed;
+    }
+}
+
+class Racer extends Horse {
+    constructor(name, breed) {
+        super(name, breed);
+        this.distance = 0;
+        this.speed = 0;
+        this.setSpeed();
+        this.time = 0;
+    }
+
+    setSpeed() { this.speed = 10 + 5 * Math.random(); }
+
+    run(timeout) {
+        const that = this;
+
+        let intervalID;
+
+        function racing() {
+            that.distance += that.speed;
+            that.time++;
+            that.setSpeed();
+            if (that.time >= timeout) clearInterval(intervalID);
+        }
+
+        intervalID = setInterval(racing, 1000);
+    }
+}
+
+class Race {
+    constructor() {
+        this.horses = [];
+    }
+
+    createRace() {
+        this.horses = horsesList.map(arr => new Racer(arr[0], arr[1]));
+    }
+
+    startRece(timeout) {
+        this.horses.forEach(obj => obj.run(timeout));
+
+        const that = this;
+
+        function racing(time) {
+            function createLogRacing(str, obj) {
+                return `${str}${obj.name}\t(${obj.breed})\t\trun\t${obj.distance} m\n`;
+            }
+
+            function logRacing() {
+                console.log(that.horses.reduce(createLogRacing, `Time = ${time} c:\n`));
+                if (time < timeout) racing(time + 2);
+                else {
+                    that.horses.sort((a, b) => b.distance - a.distance);
+                    console.log(`Winner: ${that.horses[0].name}   congratulate the champion!`);
+                }
+            }
+
+            function synchronization() {
+                if (that.horses.some(obj => obj.time < time)) setTimeout(synchronization, 16);
+                else logRacing();
+            }
+
+            setTimeout(synchronization, 2000);
+        }
+
+        racing(2);
+    }
+}
+
+const race1 = new Race();
+
+race1.createRace();
+race1.startRece(10);
```

Comment:
In general, it may work, but... I personally would never design it this way... your classes execution logics should be separated by standalone units manageable by API-like methods

`Racer` should only run, that's it... it should have method to `start()` running and `stop()` running... 

`Race` suppose to "tell" the `Racer` when to start running and when to stop, `Racer` doesn't have to manage this...

Beside this `Race` suppose to have API-like access (`startRace()` and `stopRace()` methods) to simple actions like start process and finish process, it should also have something like `logRace()` method to log current racers results

So though your solution may be acceptable, it's has a significant room for improvement and optimization :)

Reply:
Реальна логіка роботи перегонів все ж не така: гонщики виходять на доріжки й чекають сигналу старту, а коли досягають фінішу взнають самі перетнувши фінішну лінію тоді вони й зупиняються. А перегони закінчуються коли власне всі гонщики перетнуть фінішну лінію, а не за якоюсь командою.

Те що увесь виконуваний код практично виходить в одному методі не дуже добре, але я не бачу в класах іншого місця для організації власних службових функцій. Всі функції необхідні виключно для обробки даних в цьому методі доводиться в ньому ж і описувати. Як кардинально вирішити цю ситуацію я все ж не розумію.